### PR TITLE
GH-34266: [C++] Add a pivot_longer node

### LINKED ISSUE #34266: [C++] Add a pivot_longer node
### Describe the enhancement requested

Pivoting longer is an operation that is useful when a single row contains multiple observations or measurements.  For example:

| time | left_temp | right_temp |
| ---- | --------- | ---------- |
| 1    | 10        | 20         |
| 2    | 15        | 18         |

The columns `left_temp` and `right_temp` are really the combination of a "feature" (`location`) and a "measurement" (`temp`).  We can pivot longer to extract these into their own columns:

| time | location | temp |
| --- | --- | --- |
| 1    | left     | 10   |
| 1    | right    | 20   |
| 2    | left     | 15   |
| 2    | right    | 18   |

The addition of this capability should make it possible to fully bind tidyr's pivot_longer and pivot_wider (the latter can be expressed using a combination of `project` and `aggregate`).  However, there is a considerable amount of convenience that could still be added and much of this is captured in the parent issue #34265

### Component(s)

C++

### PR DESCRIPTION
**Pull request by westonpace**
* Closes: #34266

### CODE REVIEW

**Comment by lidavidm** [comment]
I can review this once I find some time to wrap my head around what pivoting exactly is

**Comment by westonpace** [comment]
> I can review this once I find some time to wrap my head around what pivoting exactly is

:laughing: If it helps, I don't know if anyone is ready to use this exactly so I'm not sure how much to rush.  I mainly did this just to annoy @ianmcook

Also, I hope the PR itself (especially the lengthy prose in `options.h`) will help explain both why and what pivot_longer is.

File: cpp/src/arrow/compute/exec/options.h [hunk 1]
```diff
@@ -582,6 +582,94 @@ class ARROW_EXPORT TableSinkNodeOptions : public ExecNodeOptions {
...
+/// | time | left_temp | right_temp |
+/// | ---- | --------- | ---------- |
+/// | 1    | 10        | 20         |
+/// | 2    | 15        | 18         |
+///
+/// The above table contains two observations per row.  There is an implicit feature
+/// "location" (left vs right) and a measurement "temp".  What we really want is:
+///
+/// | time | location | temp |
+/// | 1    | left     | 10   |
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/options.h:620]:
```suggestion
/// | time | location | temp |
/// | ---- | -------- | --- |
/// | 1    | left     | 10   |
```

**Review by lidavidm — APPROVED** [review]
The tidyverse documentation is excellent for understanding this operation! It's still awkward to wrap around because it's very flexible, but the general idea is intuitive and the implementation is fairly straightforward.

File: cpp/src/arrow/compute/exec/options.h [hunk 2]
```diff
@@ -582,6 +582,95 @@ class ARROW_EXPORT TableSinkNodeOptions : public ExecNodeOptions {
...
+/// | ---- | --------- | ---------- |
+/// | 1    | 10        | 20         |
+/// | 2    | 15        | 18         |
+///
+/// The above table contains two observations per row.  There is an implicit feature
+/// "location" (left vs right) and a measurement "temp".  What we really want is:
+///
+/// | time | location | temp |
+/// | ---- | -------- | --- |
+/// | 1    | left     | 10   |
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/options.h:621]: I typoed that, sorry

```suggestion
/// | time | location | temp |
/// | ---- | -------- | ---- |
/// | 1    | left     | 10   |
```

**Review by lidavidm — APPROVED** [review]

**Pull request merged**